public/app/features/explore/Explore.tsx

```diff
@@ -605,6 +607,23 @@ export class Explore extends PureComponent<Props, ExploreState> {
                           queryInspectorButtonActive={showQueryInspector}
                           onClickAddQueryRowButton={this.onClickAddQueryRowButton}
                           onClickQueryInspectorButton={() => setShowQueryInspector(!showQueryInspector)}
+                          onSelectQueryFromLibrary={async (query) => {
+                            if (query.datasource?.uid) {
+                              const isDifferentDatasource = !this.props.exploreActiveDS.dsToExplore
+                                .find((di) => di.datasource.uid === query.datasource?.uid)
+                                ?.exploreIds.includes(exploreId);
+                              if (isDifferentDatasource) {
+                                await this.props.changeDatasource({ exploreId, datasource: query.datasource.uid });
```

Comment:
If there already is a query this will create a mismatch and the queries won't run. As we are adding and not replacing the query I assume that can happen often. Do we want to somehow handle that here or is it something we will figure out later? If later I would at least add a comment here so nobody is confused in the meantime.

Reply:
yeah gotcha - i was gonna switch to mixed in a later PR (both in dashboards + explore), but i can do it as part of this PR 👍

Reply:
ok i've added logic to switch to mixed if there are multiple datasources present in the queries. one wrinkle is that the starting query added by dashboards is _much_ less like a query object - basically just includes the `refId` and nothing more. i've added some logic to at least make sure it includes the datasource as well, which helps bring it closer to what explore adds and keeps the behaviour matching.